File: tests/gemma4_vs_ref.c
/* Token-level greedy comparison driver: the CNET transformer forward vs an
 * external reference (llama.cpp) on the SAME GGUF bytes.
 *
 * Feeds explicit token ids (tokenization happens on the reference side —
 * CNET mines in id space), generates N greedy steps, prints the argmax
 * chain and the ordered top-K per step. tests/gemma4_vs_ref.py runs the
 * same ids through llama-server and diffs the chains; the gate is argmax
 * identity at every step. Softcap note: the forward returns UNCAPPED
 * logits; c*tanh(x/c) is monotonic so greedy argmax is unaffected.
 *
 * Usage: gemma4_vs_ref <model.gguf> <ids-csv> <steps> [topk]
 *   ids-csv like "2,9105,603" (include BOS explicitly if wanted).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "../include/cce/cce_detect.h"

#define MAX_IDS 4096

int main(int argc, char **argv) {
    cce_anymodel *am = NULL;
    cce_gguf_qwen2 *m;
    int ids[MAX_IDS];
    int n_ids = 0;
    int steps, topk = 8;
    float *logits;
    int V;
    char *csv, *tok;
    int s;

    if (argc < 4) {
        fprintf(stderr, "usage: %s <model.gguf> <ids-csv> <steps> [topk]\n",
                argv[0]);
        return 2;
    }
    csv = argv[2];
    for (tok = strtok(csv, ","); tok && n_ids < MAX_IDS;
         tok = strtok(NULL, ","))
        ids[n_ids++] = atoi(tok);
    steps = atoi(argv[3]);
    if (argc > 4) topk = atoi(argv[4]);
    if (n_ids < 1 || steps < 1 || topk < 1 || topk > 32) {
        fprintf(stderr, "bad ids/steps/topk\n");
        return 2;
    }

    if (cce_anymodel_open(&am, argv[1]) != CCE_OK || am->transformer == NULL) {
        fprintf(stderr, "cannot open transformer for %s\n", argv[1]);
        return 1;
    }
    m = am->transformer;
    V = m->vocab_size;
    logits = (float *)malloc((size_t)V * sizeof *logits);
    if (!logits) return 1;

    printf("model: layers=%d hidden=%d vocab=%d  prompt_ids=%d steps=%d\n",
           m->n_layer, m->n_embd, V, n_ids, steps);

    if (getenv("GVR_DUMP_EMB")) {
        const float *r2 = m->tok_emb.data + (size_t)2 * m->n_embd;
        int di;
        printf("tok_emb row2 [0:16]  =");
        for (di = 0; di < 16; di++) printf(" %.6f", (double)r2[di]);
        printf("\ntok_emb row2 [256:8] =");
        for (di = 256; di < 264; di++) printf(" %.6f", (double)r2[di]);
        printf("\ntok_emb row2 [3836:4]=");
        for (di = 3836; di < 3840; di++) printf(" %.6f", (double)r2[di]);
        printf("\n");
    }

    m->cur_pos = 0;
    if (cce_gguf_qwen2_forward(m, ids, n_ids, logits, V) != CCE_OK) {
        fprintf(stderr, "prompt forward REFUSED/failed\n");
        return 1;
    }

    for (s = 0; s < steps; s++) {
        /* ordered top-K, strict > tie-break keeps the earliest id (the
           oracle's convention) */
        int top[32];
        float tv[32];
        int k, j;
        for (k = 0; k < topk; k++) {
            int bi = -1;
            float bv = 0.0f;
            for (j = 0; j < V; j++) {
                int taken = 0, t2;
                for (t2 = 0; t2 < k; t2++)
                    if (top[t2] == j) taken = 1;
                if (taken) continue;
                if (bi == -1 || logits[j] > bv) { bv = logits[j]; bi = j; }
            }
            top[k] = bi;
            tv[k] = bv;
        }
        printf("STEP %3d next=%d top%d=", s, top[0], topk);
        for (k = 0; k < topk; k++)
            printf("%d:%.4f%s", top[k], (double)tv[k],
                   k + 1 < topk ? "," : "\n");

        if (s + 1 < steps) {
            int next = top[0];
            if (cce_gguf_qwen2_forward(m, &next, 1, logits, V) != CCE_OK) {
                fprintf(stderr, "step %d forward failed\n", s);
                return 1;
            }
        }
    }
    printf("DONE\n");
    return 0;
}

File: tests/gemma4_vs_ref.py
#!/usr/bin/env python3
"""gemma4 forward validation: CNET vs llama.cpp on the same GGUF.

For each prompt: tokenize via the reference server, run greedy generation on
both engines, gate on ARGMAX IDENTITY at every step. Softcap note: the
reference caps logits (30*tanh(l/30)); CNET ranks uncapped — monotone, so
argmax agrees except at fp32 saturation ties (reported, not hidden).

Usage: tests/gemma4_vs_ref.py [--server http://127.0.0.1:8083] [--steps 16]
Requires: llama-server running on the SAME GGUF; bin/gemma4_vs_ref built.
"""
import json, os, re, subprocess, sys, urllib.request

SERVER = "http://127.0.0.1:8083"
MODEL  = "/home/marble/AI/Models/gemma4-v2-Q4_K_M.gguf"
STEPS  = 16
args = sys.argv[1:]
if "--server" in args: SERVER = args[args.index("--server")+1]
if "--steps"  in args: STEPS  = int(args[args.index("--steps")+1])

PROMPTS = [
    "The capital of France is",
    "2 + 2 =",
    "The chemical symbol for gold is",
    "Roses are red, violets are",
    "def fibonacci(n):",
    "The Eiffel Tower is located in the city of",
]

def post(path, body):
    req = urllib.request.Request(SERVER + path, json.dumps(body).encode(),
                                 {"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=300) as r:
        return json.load(r)

def ref_chain(ids, steps):
    # natural EOG stop — no ignore_eos: forcing past end-of-turn suppresses
    # the true argmax and unfairly diverges from a faithful engine
    r = post("/completion", {"prompt": ids, "n_predict": steps,
                             "temperature": 0.0, "n_probs": 2})
    toks = [t["id"] for t in r.get("completion_probabilities", [])][:steps]
    gaps = []
    for t in r.get("completion_probabilities", [])[:steps]:
        tl = t.get("top_logprobs") or t.get("probs") or []
        gaps.append(round(tl[0].get("logprob", 0) - tl[1].get("logprob", 0), 4)
                    if len(tl) >= 2 else None)
    return toks, gaps

def cnet_chain(ids, steps):
    out = subprocess.run(
        ["./bin/gemma4_vs_ref", MODEL, ",".join(map(str, ids)), str(steps), "3"],
        capture_output=True, text=True, timeout=3600)
    if out.returncode != 0:
        print(out.stderr[-400:], file=sys.stderr)
        raise RuntimeError("cnet driver failed")
    chain = [int(m.group(1)) for m in
             re.finditer(r"^STEP\s+\d+ next=(\d+)", out.stdout, re.M)]
    return chain[:steps]

only = os.environ.get("GVR_ONLY")
if only:
    keys = [k.strip().lower() for k in only.split(",")]
    PROMPTS = [p for p in PROMPTS if any(k in p.lower() for k in keys)]

fails = 0
for p in PROMPTS:
    ids = post("/tokenize", {"content": p, "add_special": True})["tokens"]
    ref, gaps = ref_chain(ids, STEPS)
    got = cnet_chain(ids, STEPS)
    # compare up to the reference's natural stop (EOG ends its chain)
    n = min(len(ref), len(got))
    div = next((i for i in range(n) if ref[i] != got[i]), None)
    ok = div is None and n > 0
    fails += (not ok)
    tag = "IDENTICAL" if ok else f"DIVERGES@{div}" if div is not None else "EMPTY"
    print(f"{tag:12s} steps={n:3d}  {p!r}")
    if not ok and div is not None:
        print(f"    ref : {ref[:div+3]}   ref top1-top2 gap at divergence: {gaps[div]}")
        print(f"    cnet: {got[:div+3]}")

print(f"\n{'GATE PASSED' if fails == 0 else 'GATE FAILED'}: "
      f"{len(PROMPTS)-fails}/{len(PROMPTS)} prompts argmax-identical x {STEPS} steps")
sys.exit(1 if fails else 0)
